fuel-vm/src/interpreter/blockchain.rs

```diff
@@ -211,16 +211,37 @@ where
         c: Word,
         d: Word,
     ) -> IoResult<(), S::DataError> {
+        let mut gas_cost = self.gas_costs().ccp;
+        // Charge only for the `base` execution.
+        // We will charge for the contracts size in the `code_copy`.
+        self.gas_charge(gas_cost.base())?;
+        gas_cost.set_base(0);
```

Comment:
I know we already use this pattern where we:
1. Charge the base of the dependent cost
2. Set the base of the dependent cost to 0
3. Charge the variable amount of the dependent cost

In general, I do not think this pattern makes a lot of sense. The structure of `DependentCost` doesn't line up with how it's used: We basically split the one dependent cost into both a fixed cost and dependent cost so we can charge at different points. We break encapsulation by reading and charging `base` as a fixed cost, then we mutate `base` to 0, which means `CodeCopyCtx` (and other places we do this) requires a nonstandard mutated `DependentCost` input, but there's no type safety or documentation to let developers know. All of this breaks the principle of least surprise.

Maybe we could break up dependent costs into both a `DependentCost` and fixed cost, where dependent cost is just the variable amount and the fixed cost is the base. This way, we have type safety and no gymnastics to fit our two-stage charging pattern.

Reply:
I tried to minimise the number of changes. Updated in the https://github.com/FuelLabs/fuel-vm/pull/637/commits/7d9cfdd6a69fb8efc8043f0f723df1acd37de9ce